File: modules/gd-kasane/src/mesh_view.cpp
// SPDX-License-Identifier: MIT
#include "mesh_view.hpp"
#include <godot_cpp/core/class_db.hpp>
#include <godot_cpp/classes/os.hpp>
#include <godot_cpp/classes/rendering_server.hpp>
#include <godot_cpp/variant/packed_vector3_array.hpp>
#include <algorithm>
#include <cstring>

using namespace godot;

namespace kasane_gd {
void KasaneMeshView::_bind_methods() {
    ClassDB::bind_method(D_METHOD("initialize", "positions", "uvs", "indices", "texture"),
                         &KasaneMeshView::initialize);
    ClassDB::bind_method(D_METHOD("update_positions", "positions"), &KasaneMeshView::update_positions);
    ClassDB::bind_method(D_METHOD("clear"), &KasaneMeshView::clear);
    ClassDB::bind_method(D_METHOD("get_render_stats"), &KasaneMeshView::get_render_stats);
    ClassDB::bind_method(D_METHOD("get_positions_snapshot"), &KasaneMeshView::get_positions_snapshot);
}

Dictionary KasaneMeshView::initialize(const PackedVector2Array &positions, const PackedVector2Array &uvs,
                                      const PackedInt32Array &indices, const Ref<Texture2D> &texture) {
    if (!(OS::get_singleton()->get_thread_caller_id() == OS::get_singleton()->get_main_thread_id()))
        return error("WRONG_THREAD", "Rendering calls require the main thread.");
    if (texture.is_null() || texture->get_width() <= 0 || texture->get_height() <= 0)
        return error("INVALID_TEXTURE", "Supply a loaded Texture2D.");
    auto next = vectors(positions);
    auto next_uvs = vectors(uvs);
    std::vector<uint32_t> dense;
    for (int64_t i = 0; i < indices.size(); ++i) {
        if (indices[i] < 0)
            return error("INVALID_INDEX", "Indices cannot be negative.");
        dense.push_back(indices[i]);
    }
    if (auto status = kasane::validate_render_mesh(next, next_uvs, dense); !status.ok())
        return result(status);
    Ref<ArrayMesh> next_surface;
    next_surface.instantiate();
    PackedVector3Array gpu_positions;
    gpu_positions.resize(next.size());
    for (size_t i = 0; i < next.size(); ++i)
        gpu_positions.set(i, {next[i].x, next[i].y, 0});
    Array arrays;
    arrays.resize(Mesh::ARRAY_MAX);
    arrays[Mesh::ARRAY_VERTEX] = gpu_positions;
    arrays[Mesh::ARRAY_TEX_UV] = uvs;
    arrays[Mesh::ARRAY_INDEX] = indices;
    next_surface->add_surface_from_arrays(Mesh::PRIMITIVE_TRIANGLES, arrays, {}, {},
                                          Mesh::ARRAY_FLAG_USE_DYNAMIC_UPDATE);
    surface_ = next_surface;
    positions_ = std::move(next);
    set_mesh(surface_);
    set_texture(texture);
    set_texture_filter(CanvasItem::TEXTURE_FILTER_NEAREST);
    set_texture_repeat(CanvasItem::TEXTURE_REPEAT_DISABLED);
    update_bounds();
    ++creations_;
    return result({});
}

Dictionary KasaneMeshView::update_positions(const PackedVector2Array &positions) {
    if (!(OS::get_singleton()->get_thread_caller_id() == OS::get_singleton()->get_main_thread_id()))
        return error("WRONG_THREAD", "Rendering calls require the main thread.");
    if (surface_.is_null())
        return error("NOT_INITIALIZED", "Initialize the mesh first.");
    if (positions.size() != static_cast<int64_t>(positions_.size()))
        return error("INVALID_LENGTH", "Position updates must preserve topology and vertex count.");
    auto next = vectors(positions);
    if (auto status = kasane::validate_positions(next); !status.ok())
        return result(status);
    PackedByteArray upload;
    upload.resize(next.size() * 3 * sizeof(float));
    auto *bytes = upload.ptrw();
    for (size_t i = 0; i < next.size(); ++i) {
        const float vertex[3] = {next[i].x, next[i].y, 0};
        std::memcpy(bytes + i * sizeof(vertex), vertex, sizeof(vertex));
    }
    surface_->surface_update_vertex_region(0, 0, upload);
    positions_ = std::move(next);
    update_bounds();
    ++uploads_;
    return result({});
}

void KasaneMeshView::update_bounds() {
    float left = positions_[0].x, right = left, top = positions_[0].y, bottom = top;
    for (const auto &p : positions_) {
        left = std::min(left, p.x);
        right = std::max(right, p.x);
        top = std::min(top, p.y);
        bottom = std::max(bottom, p.y);
    }
    surface_->set_custom_aabb(AABB(Vector3(left, top, -0.5), Vector3(right - left, bottom - top, 1)));
}

void KasaneMeshView::clear() {
    ERR_FAIL_COND_MSG(
        !(OS::get_singleton()->get_thread_caller_id() == OS::get_singleton()->get_main_thread_id()),
        "Rendering calls require the main thread.");
    set_mesh(Ref<Mesh>());
    set_texture(Ref<Texture2D>());
    surface_.unref();
    positions_.clear();
}

Dictionary KasaneMeshView::get_render_stats() const {
    Dictionary out;
    out["initialized"] = surface_.is_valid();
    out["mesh_rid"] = surface_.is_valid() ? surface_->get_rid().get_id() : uint64_t(0);
    out["texture_instance_id"] = get_texture().is_valid() ? get_texture()->get_instance_id() : uint64_t(0);
    out["vertex_count"] = static_cast<int64_t>(positions_.size());
    out["position_uploads"] = uploads_;
    out["surface_creations"] = creations_;
    return out;
}
} // namespace kasane_gd

File: modules/kasane-godot/src/mesh_view.rs
use godot::classes::{
    canvas_item::{TextureFilter, TextureRepeat},
    mesh::{ArrayFormat, ArrayType, PrimitiveType},
    ArrayMesh, MeshInstance2D, Texture2D,
};
use godot::prelude::*;

use kasane_core::geometry::{validate_positions, validate_render_mesh};
use kasane_core::types::Vec2;

use crate::conversions::{
    error_dict, is_main_thread, packed_to_vectors, status_to_dict, vectors_to_packed, Array,
    Dictionary,
};

#[derive(GodotClass)]
#[class(init, base=MeshInstance2D)]
pub struct KasaneMeshView {
    base: Base<MeshInstance2D>,
    surface: Option<Gd<ArrayMesh>>,
    positions: Vec<Vec2>,
    uploads: u64,
    creations: u64,
}

#[godot_api]
impl KasaneMeshView {
    #[func]
    pub fn initialize(
        &mut self,
        positions: PackedVector2Array,
        uvs: PackedVector2Array,
        indices: PackedInt32Array,
        texture: Option<Gd<Texture2D>>,
    ) -> Dictionary {
        if !is_main_thread() {
            return error_dict("WRONG_THREAD", "Rendering calls require the main thread.");
        }
        let Some(tex) = texture else {
            return error_dict("INVALID_TEXTURE", "Supply a loaded Texture2D.");
        };
        if tex.get_width() <= 0 || tex.get_height() <= 0 {
            return error_dict("INVALID_TEXTURE", "Supply a loaded Texture2D.");
        }
        let next = packed_to_vectors(&positions);
        let next_uvs = packed_to_vectors(&uvs);
        let mut dense = Vec::with_capacity(indices.len());
        for i in 0..indices.len() {
            let idx = indices[i];
            if idx < 0 {
                return error_dict("INVALID_INDEX", "Indices cannot be negative.");
            }
            dense.push(idx as u32);
        }
        let status = validate_render_mesh(&next, &next_uvs, &dense);
        if !status.is_ok() {
            return status_to_dict(&status);
        }

        let mut next_surface = ArrayMesh::new_gd();
        let mut gpu_positions = PackedVector3Array::new();
        gpu_positions.resize(next.len());
        for (i, p) in next.iter().enumerate() {
            gpu_positions[i] = Vector3::new(p.x, p.y, 0.0);
        }
        let mut arrays = Array::new();
        arrays.resize(ArrayType::MAX.ord() as usize, &Variant::nil());
        arrays.set(
            ArrayType::VERTEX.ord() as usize,
            &gpu_positions.to_variant(),
        );
        arrays.set(ArrayType::TEX_UV.ord() as usize, &uvs.to_variant());
        arrays.set(ArrayType::INDEX.ord() as usize, &indices.to_variant());

        next_surface
            .add_surface_from_arrays_ex(PrimitiveType::TRIANGLES, &arrays)
            .flags(ArrayFormat::FLAG_USE_DYNAMIC_UPDATE)
            .done();

        self.surface = Some(next_surface.clone());
        self.positions = next;
        self.base_mut().set_mesh(&next_surface);
        self.base_mut().set_texture(&tex);
        self.base_mut().set_texture_filter(TextureFilter::NEAREST);
        self.base_mut().set_texture_repeat(TextureRepeat::DISABLED);
        self.update_bounds();
        self.creations += 1;
        status_to_dict(&kasane_core::types::Status::ok())
    }

    #[func]
    pub fn update_positions(&mut self, positions: PackedVector2Array) -> Dictionary {
        if !is_main_thread() {
            return error_dict("WRONG_THREAD", "Rendering calls require the main thread.");
        }
        if self.surface.is_none() {
            return error_dict("NOT_INITIALIZED", "Initialize the mesh first.");
        }
        if positions.len() != self.positions.len() {
            return error_dict(
                "INVALID_LENGTH",
                "Position updates must preserve topology and vertex count.",
            );
        }
        let next = packed_to_vectors(&positions);
        let status = validate_positions(&next);
        if !status.is_ok() {
            return status_to_dict(&status);
        }
        let mut upload = PackedByteArray::new();
        upload.resize(next.len() * 3 * 4);
        for (i, p) in next.iter().enumerate() {
            let offset = i * 12;
            let x_bytes = p.x.to_ne_bytes();
            let y_bytes = p.y.to_ne_bytes();
            let z_bytes = 0.0f32.to_ne_bytes();
            for b in 0..4 {
                upload[offset + b] = x_bytes[b];
                upload[offset + 4 + b] = y_bytes[b];
                upload[offset + 8 + b] = z_bytes[b];
            }
        }
        self.surface
            .as_mut()
            .unwrap()
            .surface_update_vertex_region(0, 0, &upload);
        self.positions = next;
        self.update_bounds();
        self.uploads += 1;
        status_to_dict(&kasane_core::types::Status::ok())
    }

    fn update_bounds(&mut self) {
        if self.positions.is_empty() {
            return;
        }
        let mut left = self.positions[0].x;
        let mut right = left;
        let mut top = self.positions[0].y;
        let mut bottom = top;
        for p in &self.positions {
            left = left.min(p.x);
            right = right.max(p.x);
            top = top.min(p.y);
            bottom = bottom.max(p.y);
        }
        if let Some(surf) = self.surface.as_mut() {
            surf.set_custom_aabb(Aabb::new(
                Vector3::new(left, top, -0.5),
                Vector3::new(right - left, bottom - top, 1.0),
            ));
        }
    }

    #[func]
    pub fn clear(&mut self) {
        if !is_main_thread() {
            godot_error!("Rendering calls require the main thread.");
            return;
        }
        self.surface = None;
        self.positions.clear();
    }

    #[func]
    pub fn get_render_stats(&self) -> Dictionary {
        let mut d = Dictionary::new();
        d.set("uploads", self.uploads as i64);
        d.set("creations", self.creations as i64);
        d
    }

    #[func]
    pub fn get_positions_snapshot(&self) -> PackedVector2Array {
        vectors_to_packed(&self.positions)
    }
}
